src/nlp/search.py

```diff
@@ -0,0 +1,366 @@
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples per query.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples per query.
+        """
+        assert len(queries.shape) == 2
+        if not queries.flags.c_contiguous:
+            queries = np.asarray(queries, order="C")
+        scores, indices = self.faiss_index.search(queries, k)
+        return scores, indices.astype(int)
+
+
+class IndexableMixin:
+    """Add indexing features to `nlp.Dataset`"""
+
+    def __init__(self):
+        self._indexes: Dict[str, BaseIndex] = {}
+
+    def __getitem__(self, key):
+        raise NotImplementedError
+
+    def is_index_initialized(self, column: str) -> bool:
+        return column in self._indexes
+
+    def _check_index_is_initialized(self, column: str):
+        if not self.is_index_initialized(column):
+            raise MissingIndex(
+                f"Index with column '{column}' not initialized yet. Please make sure that you call `add_vector_index` or `add_text_index` first."
+            )
+
+    def list_indexes(self) -> List[str]:
+        """List the columns/identifiers of all the attached indexes."""
+        return list(self._indexes)
+
+    def add_vector_index(
+        self,
+        column: str,
+        vectors,
+        device: Optional[int] = None,
+        string_factory: Optional[str] = None,
+        faiss_gpu_options: Optional[FaissGpuOptions] = None,
+    ):
+        """ Add a dense index using Faiss for fast retrieval.
+
+            Args:
+                `column` (`str`): The column/identifier of the index. This is the column that is used to call `.get_nearest` or `.search`.
+                `vectors` (`Union[np.ndarray, nlp.Dataset]`): The vectors to index. It can be a `nlp.Dataset` if the `dataset` is well formated
+                    using `dataset.set_format("numpy")` for example.
+                `device` (Optional `int`): If not None, this is the index of the GPU to use. By default it uses the CPU.
+                `string_factory` (Optional `str`): This is passed to the index factory Faiss to create the index. Default index class is IndexFlatIP.
+                `faiss_gpu_options` (Optional `FaissGpuOptions`): Options to configure the GPU resources of Faiss.
+                `column` (Optional `str`): In the case of a `nlp.Dataset` input, which column to index.
+        """
+        self._indexes[column] = DenseIndex(device, string_factory, faiss_gpu_options)
+        self._indexes[column].add_vectors(vectors, column=column)
+
+    def add_text_index(self, column: str, texts, es_client, index_name):
+        """ Add a text index using ElasticSearch for fast retrieval.
+
+            Args:
+                `column` (`str`): The column/identifier of the index. This is the column that is used to call `.get_nearest` or `.search`.
+                `texts` (`Union[List[str], nlp.Dataset]`): The texts to index. It can be a `nlp.Dataset`.
```

Comment:
here as well, I'm not sure I understand how the option to add new texts works with our philosophy of having all in the Arrow table.

Reply:
For vectors I can understand that there can be use cases where you'd want to use your own vectors/embeddings but not sure about text...